rakelib/plugins_docs_dependencies.rake

```diff
@@ -0,0 +1,169 @@
+    measure_execution("batch install of plugins") do
+      batch_install(plugins_to_install, successful_dependencies, failures, partition_size)
+    end
+
+    return [successful_dependencies, failures]
+  end
+
+  # If we found the result of a previous execution we will use the successful plugins result
+  # to order the current plugins, we assume that the plugin that was successful will still be successful.
+  # This help us reduce the number of resolve call and make the generation 3 times faster.
+  def optimize_for_successful_resolution(plugins_to_install)
+    if ::File.exist?(EXPORT_FILE)
+      content = JSON.parse(::File.read(EXPORT_FILE))
+
+      possible_success = []
+      possible_failures = []
+      unknown = []
+
+      plugins_to_install.each do |name|
+        if content["successful"][name].nil?
+          if content["failures"][name].nil?
+            unknown << name
+          else
+            possible_failures << name
+          end
+        else
+          possible_success << name
+        end
+      end
+
+      plugins_to_install = possible_success.concat(possible_failures).concat(unknown)
+      [plugins_to_install, plugins_to_install.size / possible_success.size]
+    else
+      [plugins_to_install, 2]
+    end
+  end
+
+  # Try to recursively do batch operation on the plugin list to reduce the number of resolution required.
+  def batch_install(plugins_to_install, successful_dependencies, failures, partition_size = 2)
+    plugins_to_install.each_slice(plugins_to_install.size /  partition_size) do |partition|
+      install(partition, successful_dependencies, failures)
+    end
+  end
+
+  def install(plugins_to_install, successful_dependencies, failures)
+    begin
+      builder = Bundler::Dsl.new
+      gemfile = LogStash::Gemfile.new(File.new(LogStash::Environment::GEMFILE_PATH, "r+")).load
+      plugins_to_install.each { |plugin_name| gemfile.update(plugin_name) }
+      builder.eval_gemfile("bundler file", gemfile.generate())
+      definition = builder.to_definition(LogStash::Environment::LOCKFILE, {})
+      definition.resolve_remotely!
+      extract_versions(definition, successful_dependencies, :missing)
+      puts "Batch install size: #{plugins_to_install.size}, Succesfully resolved: #{plugins_to_install}"
+    rescue => e
+      if plugins_to_install.size > 1
+        batch_install(plugins_to_install, successful_dependencies, failures)
+      else
+        puts "Failed to install: #{plugins_to_install.first}"
+        failures[plugins_to_install.first] = {
+          "klass" => e.class,
+          "message" => e.message
+        }
+      end
+    end
+  end
+
+  def extract_versions(definition, dependencies, from)
+    definition.specs.each do |spec|
+      dependencies[spec.name] ||= []
+      dependencies[spec.name] << VersionDependencies.new(spec.version, from)
+    end
+  end
+
+  def filter_logstash_plugins(specs)
+    specs.delete_if do |gem_name, _|
```

Comment:
We can check if the plugin is a logstash plugin using `Gem::Specification#metadata` instead of these regex tests.

Example:

```
>> Gem::Specification.all.select { |x| x.metadata["logstash_plugin"] }.first
=> #<Gem::Specification:0x2c26e logstash-codec-cef-4.1.2-java>
```

Assuming the `specs` contain Gem::Specifications, anyway.